### PR TITLE
Change DIBuilderCreateEnumerator signature to match LLVM 9

### PR DESCRIPTION
**Pull request by tmiasko**
* Change DIBuilderCreateEnumerator signature to match LLVM 9 C API. 
* Use provided is unsigned flag when emitting enumerators.

### CODE REVIEW

File: src/librustc_codegen_llvm/debuginfo/metadata.rs [hunk 1]
```diff
@@ -1779,13 +1779,20 @@ fn prepare_enum_metadata(
                 .discriminants(cx.tcx)
                 .zip(&def.variants)
                 .map(|((_, discr), v)| {
-                    let name = SmallCStr::new(&v.ident.as_str());
+                    let name = v.ident.as_str();
```
Comment (ecstatic-morse) [inline src/librustc_codegen_llvm/debuginfo/metadata.rs]:
```suggestion
                    let name: &'static str = v.ident.as_str().string;
```

`SmallCStr` is an owning type. Does `DIBuilderCreateEnumerator` take ownership of the name pointer? `as_str` returns a reference to the symbol interner, so this a use-after-free unless the interner is guaranteed to outlive our calls to LLVM. Is this the case? If we need to do away with the calls to `SmallCStr`, we should at least ensure that the lifetime of the returned `str` is indeed `static`.

File: src/librustc_codegen_llvm/debuginfo/metadata.rs [hunk 2]
```diff
@@ -1794,13 +1801,15 @@ fn prepare_enum_metadata(
                 .as_generator()
                 .variant_range(enum_def_id, cx.tcx)
                 .map(|variant_index| {
-                    let name = SmallCStr::new(&substs.as_generator().variant_name(variant_index));
+                    let name = substs.as_generator().variant_name(variant_index);
```
Comment (ecstatic-morse) [inline src/librustc_codegen_llvm/debuginfo/metadata.rs]: Same as above, does `DIBuilderCreateEnumerator` take ownership of the name pointer? If so, this is a use-after-free when `variant_name` returns the owned variant of a `Cow`. Why did you remove the `SmallCStr` here?

File: src/librustc_codegen_llvm/debuginfo/metadata.rs [hunk 3]
```diff
@@ -1794,13 +1801,15 @@ fn prepare_enum_metadata(
...
                     unsafe {
                         Some(llvm::LLVMRustDIBuilderCreateEnumerator(
                             DIB(cx),
-                            name.as_ptr(),
-                            // FIXME: what if enumeration has i128 discriminant?
-                            variant_index.as_usize() as u64,
+                            name.as_ptr().cast(),
+                            name.len(),
+                            // Generators use u32 as discriminant type.
+                            variant_index.as_usize() as i64,
```
Comment (ecstatic-morse) [inline src/librustc_codegen_llvm/debuginfo/metadata.rs]:
```suggestion
                            variant_index.index().try_into().expect("Generators must use `u32` as the discriminant type"),
```

If we're removing the FIXME, let's use `expect`.

**Review by ecstatic-morse — CHANGES_REQUESTED** [review]
Some questions about ownership need to be resolved before this is merged. Also left a nit about using `as` casts for discriminants.

File: src/librustc_codegen_llvm/debuginfo/metadata.rs (hunk 1 above)
Comment (tmiasko) [reply src/librustc_codegen_llvm/debuginfo/metadata.rs]: DIBuilderCreateEnumerator borrows the name for the duration of the call, there
is no transfer of ownership. SmallCStr was used to signal string length through
zero termination, which is now done through separate argument (which avoids
creating a temporary copy).

File: src/librustc_codegen_llvm/debuginfo/metadata.rs (hunk 3 above)
Comment (tmiasko) [reply src/librustc_codegen_llvm/debuginfo/metadata.rs]: Replaced with `as_u32().into()`

**Comment by ecstatic-morse** [comment]
Thanks @tmiasko! LGTM, although I'm gonna defer to someone who works on this code, since I don't quite understand why this worked before if we were using the wrong signature.

r? @nagisa

**Comment by tmiasko** [comment]
Oh sorry, I didn't made it clear in PR description. The PR prepares for migration
to the C API that is available since LLVM 9, without actually mgirating it yet, sice 
we support older versions of LLVM.

Thus the code continues to use own wrapper `LLVMRustDIBuilderCreateEnumerator` rather than
official one `LLVMDIBuilderCreateEnumerator`, but now they have exactly the same type
signature and arguments.

**Comment by ecstatic-morse** [comment]
Ah, I see now.

r? @ecstatic-morse 
@bors r+ rollup

**Comment by ecstatic-morse** [comment]
Note to self: `llvm::StringRef` explicitly does *not* require the string to be null-terminated. It does, however, have a `const char*` constructor for null-terminated strings, which is what we were using prior to this PR.

**Pull request merged**